### PR TITLE
feat(ec2): subnet ipv4 cidr blocks on imported vpc

### PR DESCRIPTION
**Pull request by schourode**
When using `Vpc.fromVpcAttributes()` to import a VPC, it is possible to specify all subnet properties except for the IPv4 CIDR block. This means that any attempt to read the `ipv4CidrBlock` property of a subnet on such a VPC will throw with the message:

> You cannot reference an imported Subnet's IPv4 CIDR if it was not supplied. Add the ipv4CidrBlock when importing using Subnet.fromSubnetAttributes()

This PR extends the `VpcAttributes` interface to allow for subnet IPv4 CIDR blocks to be passed in alongside the IDs, names and route table IDs for each subnet group (public, private, isolated).

I have added a unit test showing how the values passed into `Vpc.fromVpcAttributes()` end up in the correct positions across the `ImportedSubnet` instances. As far as I can tell, there are no existing integration tests for this type of import, leading me to assume an integration test is not needed for this PR either.

After checking the current README in the `@aws-cdk/aws-ec2` module, I am of the impression that nothing needs to be added here. Let me know if you believe otherwise.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Construct Runtime Dependencies:

* [ ] This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-construct-runtime-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by schourode** [comment]
@mrgrain Sorry about pinging you directly. You were helpful towards getting my previous PR into a mergeable state, so I am hoping you are able to help with this one also 🙏 

I submitted this PR early this week. It fails the PR linter due to lack of changes to README and integration tests, and I suspect the failing checks may be the reason why none from the _aws-cdk-owners_ group have looked at the PR yet.

As per the PR description, it is my opinion that neither README nor integration tests needs updating for the changes I am proposing here. If you disagree, I am happy to comply with directions on what additions you find to be needed for this PR to be acceptable.

**Comment by mrgrain** [comment]
> @mrgrain Sorry about pinging you directly. You were helpful towards getting my previous PR into a mergeable state, so I am hoping you are able to help with this one also 🙏

Hey hey, I don't mind, but FYI some really don't like this. As an explanation (not an excuse), the holiday season is coming up and for example I had very limited availability this week and am on Annual Leave from pretty much now-ish. It will be similar with other maintainers.

> I submitted this PR early this week. It fails the PR linter due to lack of changes to README and integration tests, and I suspect the failing checks may be the reason why none from the _aws-cdk-owners_ group have looked at the PR yet.

It's more that it makes the review more complicated. Now I'll have to go and check the documentation and existing tests first, taking me away from just reviewing the code. Without a review entry and without an issue attached, all we have is 3-4 paragraphs on what problem a feature is solving and how the proposed solution fits in. Sometimes that is in deed enough, but sometimes it's not and this will also vary a lot from who is looking at a PR and what their knowledge is (at the end of the day the core maintainers are CDK experts not all-around AWS experts).

I'd recommend you to always right a README entry. If you don't think it should be in the README, just put it in the PR description instead.

Generally speaking, the lack of existing integration tests does not proof that a new/changed integ test is not needed. It's about whether a change can be tested (e.g. we don't currently have to ability to effectively test updates to a resource) and if it makes sense (e.g. I've exempted a PR changing some Docker stuff recently because it would only have tested if docker works as documented).

> As per the PR description, it is my opinion that neither README nor integration tests needs updating for the changes I am proposing here. If you disagree, I am happy to comply with directions on what additions you find to be needed for this PR to be acceptable.

I disagree in this case. [This section](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2-readme.html#importing-an-existing-vpc) would very much benefit from a description on how to specify the cidr blocks and why someone would want to do this.

File: packages/@aws-cdk/aws-ec2/lib/util.ts [hunk 1]
```diff
@@ -71,6 +75,11 @@ export class ImportSubnetGroup {
       /* eslint-disable max-len */
       throw new Error(`Number of ${routeTableIdField} (${this.routeTableIds.length}) must be equal to the amount of ${idField} (${this.subnetIds.length}).`);
     }
+    if (this.ipv4CidrBlocks.length !== this.subnetIds.length && ipv4CidrBlocks != null) {
+      // We don't err if no ipv4CidrBlocks were provided to maintain backwards-compatibility. See https://github.com/aws/aws-cdk/pull/3171
+      /* eslint-disable max-len */
+      throw new Error(`Number of ${ipv4CidrBlockField} (${this.ipv4CidrBlocks.length}) must be equal to the amount of ${idField} (${this.subnetIds.length}).`);
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-ec2/lib/util.ts:81]: Add a unit test for the error cases please

File: packages/@aws-cdk/aws-ec2/lib/util.ts [hunk 2]
```diff
@@ -71,6 +75,11 @@ export class ImportSubnetGroup {
       /* eslint-disable max-len */
       throw new Error(`Number of ${routeTableIdField} (${this.routeTableIds.length}) must be equal to the amount of ${idField} (${this.subnetIds.length}).`);
     }
+    if (this.ipv4CidrBlocks.length !== this.subnetIds.length && ipv4CidrBlocks != null) {
+      // We don't err if no ipv4CidrBlocks were provided to maintain backwards-compatibility. See https://github.com/aws/aws-cdk/pull/3171
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-ec2/lib/util.ts:79]: Is that issue really related to cidr blocks?

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 3]
```diff
@@ -2084,9 +2111,9 @@ class ImportedVpc extends VpcBase {
...
-    const iso = new ImportSubnetGroup(props.isolatedSubnetIds, props.isolatedSubnetNames, props.isolatedSubnetRouteTableIds, SubnetType.PRIVATE_ISOLATED, this.availabilityZones, 'isolatedSubnetIds', 'isolatedSubnetNames', 'isolatedSubnetRouteTableIds');
+    const pub = new ImportSubnetGroup(props.publicSubnetIds, props.publicSubnetNames, props.publicSubnetRouteTableIds, props.publicSubnetIpv4CidrBlocks, SubnetType.PUBLIC, this.availabilityZones, 'publicSubnetIds', 'publicSubnetNames', 'publicSubnetRouteTableIds', 'ipv4CidrBlocks');
+    const priv = new ImportSubnetGroup(props.privateSubnetIds, props.privateSubnetNames, props.privateSubnetRouteTableIds, props.privateSubnetIpv4CidrBlocks, SubnetType.PRIVATE_WITH_EGRESS, this.availabilityZones, 'privateSubnetIds', 'privateSubnetNames', 'privateSubnetRouteTableIds', 'ipv4CidrBlocks');
+    const iso = new ImportSubnetGroup(props.isolatedSubnetIds, props.isolatedSubnetNames, props.isolatedSubnetRouteTableIds, props.isolatedSubnetIpv4CidrBlocks, SubnetType.PRIVATE_ISOLATED, this.availabilityZones, 'isolatedSubnetIds', 'isolatedSubnetNames', 'isolatedSubnetRouteTableIds', 'ipv4CidrBlocks');
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts:2116]: pretty sure the fieldname `ipv4CidrBlocks` needs to be prefixed with public, private or isolated.

**Review by mrgrain — DISMISSED** [review]

**Comment by mrgrain** [comment]
@schourode Like I've said above, I'm on leave now. I'm happy to take this PR through to its completion, but expect significant lower response times from me. If that's not okay for you (I'd understand) I'm equally happy to remove my assignment and let someone else take over (although holiday season constraints still apply). Just let me know what you'd prefer.

**Comment by schourode** [comment]
@mrgrain Thanks a million for the elaborate answer and the review. I will get to work on addressing incorrect strings, readme, tests next week. I am in no hurry to have this merged and would be happy to have you still assigned even if you won't be picking this up again until you are back at work.

When I tagged you above, it was out of a worry that the PR might fall between the cracks due to that failing lint check and would never attract any human attention. From your explanation, I understand that it would have been noticed, only this time of year it may take a while. That's completely understandable.

Again, thanks for all of the feedback here – both that concerning the proposed change and that concerning the contribution process moreover. It is really helpful.

Happy holidays 🤗

**Comment by mrgrain** [comment]
All good! You're worries are real and I can understand where they are coming from. That's something we as maintainers need to work on.

File: packages/@aws-cdk/aws-ec2/README.md [hunk 4]
```diff
@@ -527,6 +527,30 @@ const vpc = ec2.Vpc.fromVpcAttributes(this, 'VPC', {
...
+```
+
+The above example will create an `IVpc` instance with three public subnets:
+
+| Subnet id | Availability zone | Subnet name | Route table id | IPv4 CIDR   |
+| --------- | ----------------- | ----------- | -------------- | ----------- |
+| s-12345   | us-east-1a        | Subnet A    | rt-12345       | 10.0.0.0/24 |
+| s-34567   | us-east-1b        | Subnet B    | rt-34567       | 10.0.1.0/24 |
+| s-56789   | us-east-1c        | Subnet B    | rt-56789       | 10.0.2.0/24 |
+
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-ec2/README.md:553]:
```suggestion

For private or isolated subnet groups, use the properties with the respective prefix.

```

File: packages/@aws-cdk/aws-ec2/README.md [hunk 5]
```diff
@@ -527,6 +527,30 @@ const vpc = ec2.Vpc.fromVpcAttributes(this, 'VPC', {
 });
 ```
 
+For each subnet group the import function accepts optional parameters for subnet
+names, route table ids and IPv4 CIDR blocks. When supplied, the length of these
+lists are required to match the length of the list of subnet ids, allowing the
+lists to be zipped together to form `ISubnet` instances. Example:
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-ec2/README.md:533]:
```suggestion
lists to be zipped together to form `ISubnet` instances.

Public subnets example (for private or isolated subnet groups, use the properties with the respective prefix):
```

**Review by mrgrain — APPROVED** [review]
🚀

**Comment by mrgrain** [comment]
@mergifyio update
